### PR TITLE
[SYCL-MLIR]: Fix compound assign op conversion

### PR DESCRIPTION
**Pull request by etiotto**
This PR fixes the assertion described in issue https://github.com/intel/llvm/issues/7641 in comment https://github.com/intel/llvm/issues/7641#issuecomment-1341187847

A new test case has been added to prevent regressions. 

Signed-off-by: Tiotto, Ettore <ettore.tiotto@intel.com>

### CODE REVIEW

File: polygeist/tools/cgeist/Test/Verification/sycl/ex_particle_reduced.cpp [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+constexpr access::mode sycl_read_write = access::mode::read_write;
+
+void likelyhood(int Nparticles) {
+  cpu_selector dev_sel;
+  queue q(dev_sel);
+  const property_list props = property::buffer::use_host_ptr();
+  float *arrayX = (float *)calloc(Nparticles, sizeof(float));
+  buffer<float, 1> arrayX_GPU(arrayX, Nparticles, props);
+
+  /****************** L I K E L I H O O D ************************************/
```
Comment (whitneywhtsang) [inline polygeist/tools/cgeist/Test/Verification/sycl/ex_particle_reduced.cpp:31]:
```suggestion
```

File: polygeist/tools/cgeist/Test/Verification/sycl/ex_particle_reduced.cpp [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+// RUN: llvm-spirv %t.bc
+
+// Test that all referenced sycl header functions are generated.
+// RUN: llvm-dis %t.bc
+// RUN: cat %t.ll | FileCheck %s --check-prefix=LLVM --implicit-check-not="declare{{.*}}spir_func"
+
+// Test that the kernel named `kernel_parallel_for_id` is generated with the correct signature.
+// LLVM: define weak_odr spir_kernel void {{.*}}kernel_likelihood(
+// LLVM-SAME: float addrspace(1)* noundef %0, %"class.sycl::_V1::range.1"* noundef byval(%"class.sycl::_V1::range.1") align 8 %1, 
+// LLVM-SAME: %"class.sycl::_V1::range.1"* noundef byval(%"class.sycl::_V1::range.1") align 8 %2, %"class.sycl::_V1::id.1"* noundef byval(%"class.sycl::_V1::id.1") align 8 %3)
```
Comment (whitneywhtsang) [inline polygeist/tools/cgeist/Test/Verification/sycl/ex_particle_reduced.cpp:18]: Thinking that we can start to simplify the run lines. What do you think?
```suggestion
// RUN: clang++ -fsycl -fsycl-targets=spir64-unknown-unknown-syclmlir -O0 -w -c %s -o %t.O0.out 2>&1 | FileCheck %s --allow-empty --implicit-check-not="{{error|Error}}:"

// RUN: clang++ -fsycl -fsycl-device-only -O0 -w -S -emit-llvm -fsycl-targets=spir64-unknown-unknown-syclmlir %s -o - | FileCheck %s --check-prefix=LLVM

// Test that the kernel named `kernel_likelihood` is generated with the correct signature.
// LLVM: define weak_odr spir_kernel void {{.*}}kernel_likelihood(
// LLVM-SAME: float addrspace(1)* noundef %0, %"class.sycl::_V1::range.1"* noundef byval(%"class.sycl::_V1::range.1") align 8 %1, 
// LLVM-SAME: %"class.sycl::_V1::range.1"* noundef byval(%"class.sycl::_V1::range.1") align 8 %2, %"class.sycl::_V1::id.1"* noundef byval(%"class.sycl::_V1::id.1") align 8 %3)
```

File: polygeist/tools/cgeist/Test/Verification/sycl/ex_particle_reduced.cpp [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+
+#include <sycl/sycl.hpp>
+using namespace sycl;
+constexpr access::mode sycl_read_write = access::mode::read_write;
+
+void likelyhood(int Nparticles) {
+  cpu_selector dev_sel;
+  queue q(dev_sel);
```
Comment (whitneywhtsang) [inline polygeist/tools/cgeist/Test/Verification/sycl/ex_particle_reduced.cpp:26]:
```suggestion
  queue q();
```

File: polygeist/tools/cgeist/Test/Verification/sycl/ex_particle_reduced.cpp [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+  const property_list props = property::buffer::use_host_ptr();
+  float *arrayX = (float *)calloc(Nparticles, sizeof(float));
+  buffer<float, 1> arrayX_GPU(arrayX, Nparticles, props);
+
+  /****************** L I K E L I H O O D ************************************/
+  q.submit([&](handler &cgh) {
+    auto arrayX_acc = arrayX_GPU.get_access<sycl_read_write>(cgh);
+    cgh.parallel_for<class kernel_likelihood>(nd_range<1>(range<1>(10), range<1>(20)), [=](nd_item<1> item) {
```
Comment (whitneywhtsang) [inline polygeist/tools/cgeist/Test/Verification/sycl/ex_particle_reduced.cpp:34]: can probably use the simplest interface that takes a id? so we can remove line 35 as well?

Comment (etiotto) [reply polygeist/tools/cgeist/Test/Verification/sycl/ex_particle_reduced.cpp:22]: Lets leave it, is something we have no coverage for in other tests.

Comment (whitneywhtsang) [reply polygeist/tools/cgeist/Test/Verification/sycl/ex_particle_reduced.cpp:21]: We do in parallel_for.cpp, but sure.

**Review by whitneywhtsang — APPROVED** [review]

**Review by victor-eds — APPROVED** [review]

**Pull request merged**